boot/setefibootvars.go

```diff
@@ -0,0 +1,196 @@
+// and the returned boot number should be ignored.
+func findMatchingBootOption(optionData []byte) (uint16, error) {
+	variables, err := efiListVariables()
+	if err != nil {
+		return 0, err
+	}
+	bootVarRegexp, err := regexp.Compile("^Boot[0-9A-F]{4}$")
+	if err != nil {
+		// error should never occur
+		return 0, err
+	}
+	usedBootNums := make(map[uint64]bool)
+	for _, varDesc := range variables {
+		varName := varDesc.Name
+		varGUID := varDesc.GUID
+		if !bootVarRegexp.MatchString(varName) {
+			// Not a Boot#### variable, so skip it
+			continue
+		}
+		if varGUID != efi.GlobalVariable {
+			// Not an EFI global variable, so skip it
+			continue
+		}
+		varNumber, err := strconv.ParseUint(varName[4:], 16, 16)
+		if err != nil {
+			// Should not occur, since variable matched bootVarRegexp
+			return 0, err
+		}
+		// Since we never overwrite an existing variable, we can ignore
+		// variable attributes when reading the variable
+		varData, _, err := efiReadVariable(varName, varGUID)
+		if err != nil {
+			return 0, err
+		}
+		if reflect.DeepEqual(optionData, varData) {
+			// existing variable already identical, use it
+			return uint16(varNumber), nil
+		}
+		usedBootNums[varNumber] = true
+	}
+	for bootNum := uint64(0); bootNum <= 0xFFFF; bootNum++ {
+		if usedBootNums[bootNum] == false {
+			return uint16(bootNum), ErrNoMatchingVariable
+		}
+	}
+	return 0, ErrAllBootNumsUsed
+}
+
+// Ensures that a Boot#### variable contains the given EFI load option.
+//
+// It may be the case than an existing boot variable already contains the
+// given load option, in which case that boot variable is reused. Otherwise,
+// finds the first unused boot variable number and uses it. Writes the load
+// option to that variable, and returns the number of the variable that was
+// used.
+func SetEfiBootOptionVariable(loadOptionData []byte) (uint16, error) {
+	bootNum, err := findMatchingBootOption(loadOptionData)
+	if err == nil {
+		return bootNum, nil
+	} else if err != ErrNoMatchingVariable {
+		return 0, err
+	}
+	varName := fmt.Sprintf("Boot%04X", bootNum)
+	err = efiWriteVariable(varName, efi.GlobalVariable, defaultVarAttrs, loadOptionData)
+	return bootNum, err
+}
+
+// Reads the current BootOrder variable, inserts the given newBootNum at the
+// beginning of the number list (and removes it from later in the list if
+// it occurs) and writes the list as the new BootOrder variable.
+func SetEfiBootOrderVariable(newBootNum uint16) error {
+	origData, attrs, err := efiReadVariable("BootOrder", efi.GlobalVariable)
+	if err == efi.ErrVarNotExist {
+		attrs = defaultVarAttrs
+		origData = make([]byte, 0)
+	} else if err != nil {
+		return err
+	}
+	if len(origData)%2 != 0 {
+		return ErrInvalidBootOrder
+	}
+	var optionOffset int
+	for optionOffset = 0; optionOffset < len(origData); optionOffset += 2 {
+		bootNum := binary.LittleEndian.Uint16(origData[optionOffset:])
+		if newBootNum == bootNum {
```

Comment:
If there is a coincidence for the first in the list, this should just return to avoid an unneeded write in each invocation.